File: Problem152.hs
module Problem152 where

import Data.List
import Data.MemoCombinators
import qualified Data.Numbers.Primes as P

dim :: Integer
dim = 80

small n = n <= dim `div` 4

primeFactors = arrayRange (2,dim) P.primeFactors

numbers = filter (all small . primeFactors) [2 .. dim]

denominator = (foldl lcm 1 numbers)^2

k |? n = n `mod` k == 0

choices = memo2 bits bits choices'
    where
    f 2 = (denominator ==) . (2*)
    f p = let p2m = (last . takeWhile (<= dim) $ iterate (p*) 1)^2
          in  (p2m |?)
    choices' p acc = filter (f p) . map (sum . (acc:)) . subsequences .
                        map ((denominator `div`) . (^2)) . filter (\x -> p |? x && all (p >=) (primeFactors x)) $ numbers
        

solution = foldr (=<<) (return 0) (map choices $ takeWhile small P.primes)

main = print $ length solution

File: problem152.cc
// problem152.cc

#include <iostream>
#include <vector>

using namespace std;

const unsigned NUMBERS[] = { 2,3,4,5,6,7,8,9,10,12,14,15,16,18,20,21,24,25,27,28,30,32,35,36,40,42,45,48,49,50,54,56,60,63,64,70,72,75,80 };
const unsigned DENOM = 2116800;
    
int main() {
    vector<unsigned> memory(DENOM/2+1, 0);
    memory[0] = 1;

    
}